LOG-545: Cluster logging Elasticsearch rollover data design

jcantrill opened the pull request:
Ref: https://jira.coreos.com/browse/LOG-545

cc  @bparees  @ewolinetz @richm @lukas-vlcek @igor-karpukhin @portante @pavolloffay @billmvt

abhinavdahiya commented:
/uncc @abhinavdahiya

portante commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,250 @@
+
+## Summary
+
+This proposal alters the data design for storing logs in ElasticSearch to co-locate logs
```
s/ElasticSearch/Elasticsearch/

portante commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,250 @@
+## Summary
+
+This proposal alters the data design for storing logs in ElasticSearch to co-locate logs
+to fewer indices.  It additionally leverages the [Elasticsearch rollover api](https://www.elastic.co/guide/en/elasticsearch/reference/6.8/indices-rollover-index.html) to
```
Since v6 of Elasticsearch is rapidly running to end-of-life, can we focus this work on Elasticsearch v7?

jcantrill replied:
No.  Only because there is no simple upgrade path from 5 to 7.  We would have to do some trickery to migrate to 6 then 7 which requires us to package both versions and handle it in the operator. More complicated then we have resources to address.... as much as we would prefer to just go straight to 7

portante replied:
Why not use the reindex [API](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-reindex.html)?

jcantrill replied:
Landed on:
* aliasing the previous indices and depending on DSL to properly secure 
* Allow previous curation to remove old indices over time

portante commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,250 @@
+
+## Motivation
+
+The initial data design for Cluster Logging segments logs by OpenShift namespace in order to facilitate multi-tenant support and data curation. This means at any one time, there are at least "**$noOfNamespaces * $daysRetatined**" number of indices maintained by the ElasticSearch server.  Each index can additionally be sharded to spread the load across the ElasticSearch nodes.  The end
```
Should we mention that this was due to SearchGuard only offering index level security in the open-source code?

Also, s/ElasticSearch/Elasticsearch/.

portante commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,250 @@
+## Motivation
+
+The initial data design for Cluster Logging segments logs by OpenShift namespace in order to facilitate multi-tenant support and data curation. This means at any one time, there are at least "**$noOfNamespaces * $daysRetatined**" number of indices maintained by the ElasticSearch server.  Each index can additionally be sharded to spread the load across the ElasticSearch nodes.  The end
+result of this shard explosion is the ElasticSearch cluster performance is not optimized and is not capable of efficiently processing logs or storing as many as it could. 
```
s/ElasticSearch/Elasticsearch/

portante commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,250 @@
+
+
+The goals of this proposals are:
+* Utilize a data design that aligns data schema with ElasticSearch's recommendations.
```
s/ElasticSearch/Elasticsearch/

portante commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,250 @@
+### Non-Goals
+
+This change will not:
+* Provide a general data management policy API that fully exposes ElasticSearch's rollover API
```
s/ElasticSearch/Elasticsearch/

portante commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,250 @@
+
+* ClusterLogging will expose the minimal needed set of ElasticSearch rollover policy management API in order to achieve the previously described goals
+* ClusterLogging will manage rollover as ElasticSearch index management is either restricted by Elastic licensing or not available in the opensource version of OpenDistro
+* Security will be addressed by using the OpenDistro security plugin. Details TBD
```
At least should we mention that Document Level Security will be required to do this?

portante commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,250 @@
+
+#### Assumptions
+
+* ClusterLogging will expose the minimal needed set of ElasticSearch rollover policy management API in order to achieve the previously described goals
```
s/ElasticSearch/Elasticsearch/

portante commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,250 @@
+#### Assumptions
+
+* ClusterLogging will expose the minimal needed set of ElasticSearch rollover policy management API in order to achieve the previously described goals
+* ClusterLogging will manage rollover as ElasticSearch index management is either restricted by Elastic licensing or not available in the opensource version of OpenDistro
```
s/ElasticSearch/Elasticsearch/

portante commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
````diff
@@ -0,0 +1,250 @@
+```
+#### Deprecations
+* The Curator Cronjob deployed by the `cluster-logging-operator` will be removed. The responsibilities for curation will be subsumed by
+ implementation of ElasticSearch rollover management.
````
s/ElasticSearch/Elasticsearch/

portante commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,250 @@
+
+## References
+
+* [ElasticSearch Rollover API documentation](https://www.elastic.co/guide/en/elasticsearch/reference/6.8/indices-rollover-index.html)
```
s/ElasticSearch/Elasticsearch/

jcantrill commented:
/hold

ewolinetz commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
````diff
@@ -0,0 +1,253 @@
+
+```
+#### Cluster Logging Operator
+The `cluster-logging-operator` will utilized the ClusterLogging CR retention policy to spec
````
s/utilized/utilize/

ewolinetz commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,253 @@
+  name: "instance"
+spec:
+  logStore:
+    retentionPolicy: 
```
i think since we talked about just exposing the elasticsearch api spec at this level, it should look the same for both EO and CLO. 

We do this currently for things like our elasticsearch storage: https://github.com/openshift/cluster-logging-operator/blob/master/pkg/apis/logging/v1/clusterlogging_types.go#L62

jcantrill replied:
See https://github.com/openshift/enhancements/pull/108/files#r346464318

jcantrill replied:
Per our architecture discussion and the desire to ensure consumers like jaeger or other can control policy, CLO is a high level abstraction that provides minimal exposure to admins.  CLO will produce a spec for the EO.  Rollover policy in EO should be flexible enough and similar for almost any deployment of Elasticsearch

ewolinetz commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,253 @@
+              max_age:   3d
+        delete:
+          min_age: 7d
+    templates:
```
i think we should abstract this slightly... this is a lot of knobs. maybe we can tuck this under something similar to what you have for `retentionPolicy` in the ClusterLogging api

jcantrill replied:
> i think we should abstract this slightly... this is a lot of knobs. maybe we can tuck this under something similar to what you have for `retentionPolicy` in the ClusterLogging api

We should have a larger discussion here with respect to who is defining what?  I understand the desire to hide this behind API but the general logic from the beginning has been: CLO provides abstract, EO allows more find tuning.  Given there is an opportunity to apply these performance improvents to any ES deployment we likely do not want to have CLO and Jaeger have to specify the same thing.  We do however, at a minimum need a way for the parent operator to tell the child to which indices we apply these changes.  This will allow a more flexible us of Elasticsearch by cluster logging and jaeger which is where we desire to go.  

I'm open to suggestions but I don't think there is anyway for EO and CLO to share the same retention policy API

jcantrill replied:
Simplified the api with the assumption the EO will infer the templates and indices to create

ewolinetz commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,253 @@
+* Create and seed specified index templates
+* Create the initial indices as needed
+* Block data ingestion if needed until initial index is seeded
+* Deploy curation CronJob for each entry in the specified seedings to rollover using the defined policy
```
EO will be responsible for this now?

ewolinetz replied:
We probably can get away with a single curator cronjob, not sure we need one per entry

jcantrill replied:
> We probably can get away with a single curator cronjob, not sure we need one per entry

I believe we can greatly simply our curation needs.  Do we really need the actual curator python code to perform this function?  We should be able to reduce it to something like a `curl` statement which has a singular responsibility of rolling a specific index which means one per.  Doesn't seem the "load" is significant

richm replied:
> Do we really need the actual curator python code to perform this function?

no

> We should be able to reduce it to something like a curl statement

or the golang elasticsearch api, or even a generic REST api

ewolinetz commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,253 @@
+#### Upgrade
+The `elasticsearch-operator` will migrate existing indices to the new data design:
+* Indices beginning with `project.*` are indexed to `container.app-write`
+* Indices beginning with `.operations.*` are indexed to `journal.infra-write`
```
is this because we currently co-locate our node and container infra logs? e.g. we just are trying to prevent losing logs with this (given we will be writing to a `node.` or `container.` for infra)

jcantrill replied:
Correct so we do not lose existing logs. It's possible we could actually spin through the `.operations` index and split out the journal and container logs but I imagine it is easier to move the index whole sale instead of trying to split it.   Opinions?

portante replied:
Why have the distinction of container vs journal logs to start?

jcantrill replied:
Only reason it exists because node logs come from journal.  My original thought was to call it `node.XXXX`... would that be better?

portante replied:
Perhaps so that it is not tied to a name referring to the implementation of how logs from the node are collected.

jcantrill replied:
replaced 'journal' with 'node'

ewolinetz commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,253 @@
+**Note:** Migrated container infrastructure logs are available via the `infra` alias only and
+subsequently curated from the cluster during period rollover.
+
+The `cluster-logging-operator` will remove previously deployed CronJobs
```
maybe specifically call out it'll remove ones that it previously owned only.

bparees commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,253 @@
+of indices maintained by the Elasticsearch server.  Each index can additionally be sharded to spread the load across the Elasticsearch nodes.  The end
+result of this shard explosion is the Elasticsearch cluster performance is not optimized and is not capable of efficiently processing logs or storing as many as it could. 
+
+Each index adds load and overhead (e.g. mapping, metadata) to the Elasticsearch cluster that needs to be tracked beyond the actual data.   Elasticsearch has recommendations for maximum shard size and the number of shards per node per allocated gigabyte of heap.  Cluster logging typically exceeds these recommendations for any OpenShift clusters that has significant log traffic.
```
do we exceed it in shard size, or shard count?  reducing the number of indexes would, i assume, increase the size of the shards, right?

portante replied:
Yes, which is where Elasticsearch is most efficient and its design point.

bparees replied:
ok.  the statements in this EP says "ES has recommendations for maximum shard size + number of shards per node.  cluster logging typically exceeds these recommendations..."

So if we're already exceeding the shard size recommendation, doing something that would make the shards bigger would seem problematic.  So is the statement incorrect?  Are we *not* currently exceeding shard size recommendations?  Or is there something else i'm missing here?

portante replied:
The new rollover scheme prevents us from exceeding the size recommendations, and keeps the number of shards down by leveraging Document Level Security to store all logs from different namespaces in the same set of indices.

So today, we have 3 shards per index (3 ES pods), one index per namespace per day.  So for 7 day retention of 100 namespaces, we have 3 * 7 * 100, or 2,100 shards.  We don't put an upper bound on data per shard, so heavy logging traffic can grow shards above the recommended size.

With the new scheme, with 7 day retention, if all the data in a day is less than the recommended size, we'll have 3 * 7 shards, 21 shards.  If the data gathered in a day is 3 times the recommended size, then we'll have 63 shards.  If it is 10 times the recommended size, we'll have 210 shards.  All significantly less than the current scheme and must more efficient because the HEAP used to track all the shards is reduced.

bparees replied:
ok, i missed the part where we are going to cap the shard size and introduce new shards when the size is reached.

So there is no risk of us exceeding recommended shard *size*.

We may still exceed recommended shard count, depending on configured retention limits and data logging rates.

portante replied:
Possibly that is why we need https://github.com/containers/conmon/issues/84 accepted in to `conmon` and then by CRI so that Kubernetes can be plummed to give somebody a hope of managing byte-capture of `stdout`/`stderr`.

bparees commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,253 @@
+
+The goals of this proposals are:
+* Utilize a data design that aligns data schema with Elasticsearch's recommendations.
+* Expose data management policy as API in the `cluster-logging-operator` and `elasticsearch-operator` in support of Cluster Logging's mission to gather container and infrastructure logs
```
is it a goal or non-goal to migrate existing customer log data to this new design?

portante replied:
Seems like if this is implemented with the v7 rollout of Elasticsearch (okay, v6), then migrating from v5 to v6 could take care of moving old to new.

bparees replied:
it was a bit of a leading question but yes i expect data migration to be a goal.  but it should be stated clearly (or indicated that it is not)

portante replied:
If we don't migrate from 5 to 7 now, then we'll have to have another release real soon that will migrate from 6 to 7 since 6 is going to be EOL "soon".

We have to solve the data migration no matter what, one way or another.  It would be great if this plan laid out how that migration is going to be handled.

jcantrill replied:
Added migration goal with additional details in the implementation section that follows

bparees commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,253 @@
+
+Logs of a given type (e.g. app container, infra container, node) are separated by index.  The Cluster Logging collector writes logs to a well-known alias established by the `cluster-logging-operator`.  The ClusterLogging CR instance specifies the management policy for each log type index.  This controls: 
+* maximum number of logs per index (e.g. 30G) 
+* the maximum age (e.g. 7 days).  
```
how does this retention policy compare w/ the current defaults?  will logs be deleted that previously were retained?

portante replied:
If you have different retention policies per namespace, then it seems like we'd need indices per class of retention policy.  So indices for 7 day retention, 6 day retention, etc.  It seems like the index size is fixed by the administrator, so not a consideration.

jcantrill replied:
It is possible logs could be deleted if they get rolled over in the index.  The best I can tell rolling over using 'max_age' is based on the index age and not documents it contains.

We do not intend to have different retention policies per namespace.  I believe you could do this in 3.x and technically in 4.x but in 4.x we never exposed API for per namespace retention policy.  This design should be extendable to allow separating logs into separate indices but that is not a goal for this proposal.

jcantrill replied:
updated verbiage to remove log size because it will be controlled by  ES recommendations

bparees commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,253 @@
+|---------|-----------|---------------|---------------|
+| Node (`logs.infra`)|infra, infra.journal|infra.journal-write|infra.journal-00001|
+| Container Infra (`logs.infra`)|infra, infra.container|infra.container-write|infra.container-000001|
+| Container Application (`logs.app`)|app.logs|app.container-write|app.container-000001|
```
where do audit logs go?

jcantrill replied:
Is it reasonable that audit logs should be separated from infra?

bparees replied:
i think so.....we're treating them separately for forwarding, right?

JAORMX replied:
It certainly is reasonable. They tend to have different usage and even data retention policy than infra logs.

jcantrill replied:
added audit logs to data model

bparees commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,253 @@
+      logs.app:
+        maxAge: 7d
+      logs.infra
+        maxAge: 7d
```
it seems awkward to me that we have retention policies at both the elastic + clusterlogging config levels..what happens when they conflict w/ each other?

jcantrill replied:
> it seems awkward to me that we have retention policies at both the elastic + clusterlogging config levels..what happens when they conflict w/ each other?

by both do you mean we express something on CL and we express something on 'Elastic'.  As mentioned here: https://github.com/openshift/enhancements/pull/108/files#r346466953 CLO design goal has been to be a higher level abstraction where Elasticsearch allows more 'knob' tuning.  All of the "Elasticsearch" settings would be generated by CLO so its fundamentally not "public" to consumers of cluster logging

bparees replied:
yeah that's what i meant, but if you're saying that the user/admin will not be allowed to edit the EO config settings (or they will be stomped by the CLO if they do) then that's fine.

ewolinetz replied:
to close the loop on this, EO is the final source of truth when it comes to policing the actual Elasticsearch cluster.
CLO (when it's CR is in a `managed` state) will end up enforcing that the elasticsearch CR spec matches what it has in its own CR.

bparees commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,253 @@
+
+## Drawbacks
+
+The drawback to note implementing this change is that Cluster Logging will:
```
did you mean 

```suggestion
The drawback to not implementing this change is that Cluster Logging will:
```

?

but normally this section is for drawbacks associated w/ implementing it.

portante commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,253 @@
+The `elasticsearch-operator` will migrate existing indices to the new data design:
+* Indices beginning with `project.*` are indexed to `container.app-write`
+* Indices beginning with `.operations.*` are indexed to `journal.infra-write`
+* Migrated indices are deleted after migration
```
How does the migration process handle the situation where the retention policy is changed along with the upgrade?

jcantrill replied:
> How does the migration process handle the situation where the retention policy is changed along with the upgrade?
The data will be subject to the new retention policy.  It's possible we could infer their policy from the previous curator deployment but I don't know how that manifests itself into the CL instance since the operator is not supposed to modify the spec.  @bparees ?

bparees replied:
reducing their retention on upgrade(and thus immediately throwing away some set of logs) seems like a non-starter.

portante replied:
So, if I have a lower retention policy in the new cluster, when the data is migrated, is it immediately deleted?  If it is not, is that a problem?

jcantrill replied:
We could additionally handle old data retention and curation by simply doing nothing, including not migrating data to the new schema.  We would leave the existing cronjob and ES would be optimized over time as old data is curated away to the point where data in the old schema is no longer present.  This would be an easier path for us to follow and addresses @portante concerns.  This will work as long as curator does not match any indices from the new schema which I believe it does not.

jcantrill replied:
@bparees @sichvoge any reason not to simply skip data migration and allow curation to curate them out. it would mean we we have to make adjustments to view the data but we would avoid the need to move data.  most of our dedicated clusters restrict data to 7 days now because of scale issues

bparees replied:
is curation on by default?  or might some clusters not have any curation happening?

if we can tolerate both data schemas from a query/view perspective, indefinitely, then this seems like an ideal answer regardless...

jcantrill replied:
Landed on leaving aliases alone and aliasing them to the new format, relying on DSL for security, and curation to roll off.

ewolinetz replied:
> Landed on leaving aliases alone and aliasing them to the new format, relying on DSL for security, and curation to roll off.

leaving **pre-existing indicies** alone and adding them to the new alias format.

jcantrill commented:
cc @alanconway

richm commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,253 @@
+status: implementable
+---
+
+# Cluster Logging Elasticsearch Rollover Data Design
```
It isn't really about just rollover data design - it is about using rollover + single indices in conjunction with DLS for access control/multi-tenancy

jcantrill replied:
This proposal is strictly about modifying the data design and introducing rollover.  It is correct that in order for us to full-fill our multitenant requirements we will need to implement DLS but that is to be addressed in a separate proposal. I would expect we do both or neither.

portante replied:
Seems like you need to explicitly link this design to be depend on the DLS design being accepted, otherwise this design is a non-starter.

jcantrill replied:
@portante that is not correct.  Rollover addresses our scalability and performance concerns but not multi-tenancy requirements.  It's possible that we could design something that sits in front of ES to restrict access which is orthogonal to roll over.

lukas-vlcek replied:
> I would expect we do both or neither.

Even without DLS the roll-over would be still great improvement. If DLS is not in place we should still push for use of roll-over.

portante replied:
> @portante that is not correct. Rollover addresses our scalability and performance concerns but not multi-tenancy requirements. It's possible that we could design something that sits in front of ES to restrict access which is orthogonal to roll over.

If you have N namespaces, you still have N indices if the size of the data in those indices does not meet the rollover limits.  The scale of the number of indices is still based on namespaces, not size/time.

So this might alleviate the issue in some cases, but without DLS, or some other implementation that addresses multi-tenancy, it is not clear how this can be a "design" that would be accepted and implemented in the product.

pavolloffay commented:
I have concerns about handling or exposing rollover configuration in Elasticsearch CR. Handling rollover seems to me like application-specific (CLO) task and not elasticsearch concern. 

Consider making the retention part in ES CR more generic (to include alias name) or just manage the rollover only in CLO CR. 

Just for completeness, I will describe how we have implemented rollover in Jaeger. The retention policies (size, number of days) can be specified in Jaeger CR. Based on the provided configuration Jaeger operator creates a job to initialize storage (create aliases, install index template) and then two cron jobs. The first one calls rollover API and the second one removes an index from the read alias.

jcantrill commented:
> I have concerns about handling or exposing rollover configuration in Elasticsearch CR. Handling rollover seems to me like application-specific (CLO) task and not elasticsearch concern.
> 

I disagree with this statement given that rollover addresses the scalability and performance of Elasticsearch.  It is a management API to allow ES indices to be curated based on policy.  The policy should be flexible enough for an application to define and the EO to manage.  There is no reason for the management to be replicated by multiple higher level components.  I would point out Elastic will provide this management for you if you use the licensed version.  Additionally, this proposal would take advantage of the OpenDistro index management if it existed in the ES6 version of their product.

> Consider making the retention part in ES CR more generic (to include alias name) or just manage the rollover only in CLO CR.

CL is an abstraction of logging and not Elasticsearch.  The policy will reflect data retention regardless of the implemented technology.  CLO will not be controlling rollover in order to better isolate the separation of concerns

> 
> Just for completeness, I will describe how we have implemented rollover in Jaeger. The retention policies (size, number of days) can be specified in Jaeger CR. Based on the provided configuration Jaeger operator creates a job to initialize storage (create aliases, install index template) and then two cron jobs. The first one calls rollover API and the second one removes an index from the read alias.

My understanding of Jaeger is that it is possible to use alternate storage engines which sounds like these policies don't apply to anything but ES.  It's likely the proposed changes are made to EO that you can remove the same functionality in the Jaeger Operator.  Implemented correctly, there is no reason you could still not use Jaeger rollover by simply not defining indexManagement on the Elasticsearch resource.

lukas-vlcek commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,246 @@
+* The Curator Cronjob deployed by the `cluster-logging-operator` will be removed. The responsibilities for curation will be subsumed by
+ implementation of Elasticsearch rollover management.
+* Curation configuration by namespace is no longer configurable and is restricted to cluster
+wide settings associated with log type 
```
I am not a fan of not having a possibility to control amount of logs per project. It is fair to say that it comes with risks.

Say some developer turns on `TRACE` level or makes unwise decision (or simply a mistake) and one or a few projects start generating a lot of logs. As of now it is at least possible to go and ad-hoc delete individual history indices for project. But when all logs are blend into a common index there is no such possibility and all other projects will get hit as well (even whole ES cluster can easily suffer) with no simple solution (except from curating all the common logs for projects sharing this index).

If we can not control this on a project level then we should at least make sure we ship appropriate alert rules to inform cluster admins that logs suddenly started growing much faster and something needs to be done about this.

portante replied:
This is the reason why we need to node-level controls for logging rates, see https://github.com/containers/conmon/issues/84

jcantrill replied:
@lukas-vlcek This change only would deprecate namespace curation for now.  We have discussed modifying LogForwarding to allow include/exclude which maybe its possible to also address throttling.

lukas-vlcek commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
````diff
@@ -0,0 +1,246 @@
+    "app.logs": {}
+   }
+}'
+```
````
I just assume the `.all` alias will be still available, right?
It is useful for Kibana navigation when redirecting from OpenShift Web console, see [this Jira comment](https://jira.coreos.com/browse/LOG-452?focusedCommentId=149472&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-149472).

portante replied:
Is this a public review of a design document that others outside of Red Hat can contribute to?  If so, then an internal Jira URL seems to be not as "open" as we should about this source.  If though, this is really just for Red Hat internal, then not sure why we are even tracking this document publicly (at least this repo appears to be public).

jcantrill replied:
> Is this a public review of a design document that others outside of Red Hat can contribute to? If so, then an internal Jira URL seems to be not as "open" as we should about this source. If though, this is really just for Red Hat internal, then not sure why we are even tracking this document publicly (at least this repo appears to be public).

This is a public forum to address and disseminate information.

portante commented:
> Just for completeness, I will describe how we have implemented rollover in Jaeger. The retention policies (size, number of days) can be specified in Jaeger CR. Based on the provided configuration Jaeger operator creates a job to initialize storage (create aliases, install index template) and then two cron jobs. The first one calls rollover API and the second one removes an index from the read alias.

Does the Jaeger operator work with a separate Elasticsearch operator?

jcantrill commented:
> Does the Jaeger operator work with a separate Elasticsearch operator?

It is the same operator which watches globally.  There is no way to deploy multiple versions of the CRD on the cluster, or at least there was not when we originally released it.

jcantrill commented:
> Regarding which CRs should handle roll-over: Are we building the Elasticsearch CR to some more abstract notion of storage management that allows a consumer of that abstraction to specify constraints for how it operates? If so, then that it would make sense to hide the roll-over implementation detail behind that abstraction.
> 

The general guiding principle is CL is abstract and ES allows us to tune more knobs 


> But if the consumer still has to "know" it is Elasticsearch under the covers, then it seems like retention should be specified once in the CL CR since the abstraction still leaks details about its implementation.

There is no reason a consumer of cluster logging needs to know we are using Elasticsearch and thus the separation.

portante commented:
> There is no reason a consumer of cluster logging needs to know we are using Elasticsearch and thus the separation.

I meant by "consumer" not the SRE team deploying cluster logging, but the difference between the CL CR and ES CR.  The CL CR in a sense "depends" on the ES CR.  So if the ES CR is not abstract, then the notions it offers must leak into how the CL CR is constructed.  Which one could argue means it is not worth having the retention related configuration in both places.

pavolloffay commented:
> My understanding of Jaeger is that it is possible to use alternate storage engines which sounds like these policies don't apply to anything but ES. It's likely the proposed changes are made to EO that you can remove the same functionality in the Jaeger Operator. Implemented correctly, there is no reason you could still not use Jaeger rollover by simply not defining indexManagement on the Elasticsearch resource.

In that case the index management API in the elastic CR should allow configuring the alias name which will be used for rollover. My impression was that it will allow to rollover only aliases used by cluster logging.


> Does the Jaeger operator work with a separate Elasticsearch operator?

It's tightly integrated with ECL ES operator, if one wants to use elatic.co operator all the configuration have to be provided explicitly https://www.jaegertracing.io/docs/1.15/operator/#external-elasticsearch

jcantrill commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
````diff
@@ -0,0 +1,242 @@
+the desired aforementioned Elasticsearch CR indexManagement.
+#### Elasticsearch API
+```
+apiVersion: "elasticsearch.openshift.io/v1alpha1"
````
@ewolinetz moving our discussion here:
```Eric Wolinetz  16 hours ago
@jcantril i'm still not sold on it being a separate CR.. is that based on https://github.com/openshift/enhancements/pull/108/files#r346559159 ?


Eric Wolinetz  16 hours ago
we could make the elasticsearch level CR be the source of truth.. this way jaeger could specify it on their CR and then CLO could also specify it for EO to consume...```

The advantages of one instance per:
* the instance explicitly defines management for one index which is specific to that index
* Index Management is not dependent on ES instance other then a ref to which specific instance (not reflected here)
* Index Management changes can be processed without needing to evaluate ES changes
Disadvantages:
* Need to explicitly ref the Elasticsearch instance to gather secrets, and ES service
* Handle case where ES doesnt exist

@bparees @alanconway Do you have an opinion of handling in this way versus previous commits where it was added to the ES resource?

ewolinetz commented on enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md:
```diff
@@ -0,0 +1,246 @@
+
+#### Assumptions
+
+* ClusterLogging will expose the minimal needed set of Elasticsearch rollover policy management API in order to achieve the previously described goals
```
can we clarify which "Elasticsearch" we're referring to here? I'm assuming its the `elasticsearch` CRD and not the component Elasticsearch...

jcantrill replied:
This line "Elasticsearch rollover policy management API" I am referring  to the actual Elasticsearch rollover api.

bparees commented:
@jcantrill this lgtm.

jcantrill commented:
/hold cancel

ewolinetz commented:
/lgtm
